lib/randomfileaccess.js

```diff
@@ -152,20 +152,24 @@ class RandomFileAccess {
      * 
      * @param {String|URL} path Path of the file to read from
      * @param {SubRange} range Range of bytes to read
-     * @returns {Buffer} Buffer of bytes read
+     * @returns {TransferMemoryBlock} Transfer memory block containing a buffer of bytes read
      */
     async read(path, range) {
         const { fileHandles } = this[PRIVATE];
 
-        console.log(`Getting file handle for reading file ${path}`);
+        const bufferMemoryBlock = await this[PRIVATE].bufferMemoryAllocator.obtainBuffer(range.length);
         const handle = await getOrOpen(fileHandles, path, "r");
+        await handle.read(bufferMemoryBlock.buffer, 0, range.length, range.low);
+        
+        return bufferMemoryBlock;
+    }
 
-        const buffer = Buffer.allocUnsafe(range.length);
-        console.log(`Allocating ${range.length} for reading from file ${path}`);
-
-        const { bytesRead } = await handle.read(buffer, 0, range.length, range.low);
-        console.log(`Read ${bytesRead} from file ${path}`);
-        return buffer.slice(0, bytesRead);
+    /**
+     * Releases memory held by a buffer used to read from a file
+     * @param {TransferMemoryBlock} bufferMemoryBlock 
+     */
+    releaseFileBufferMemory(bufferMemoryBlock){
+        this[PRIVATE].bufferMemoryAllocator.releaseBuffer(bufferMemoryBlock);
     }
```

Comment:
I think it would be better if the buffer is passed in. The range can easily provide the length, and then `read` just has to make sure that the passed in buffer is at least the length requested in the range.

That way you don't need to wrap the `releaseBuffer` function.

Reply:
I don't think this is a good idea, because it externalizes the buffer handling again. I think all buffer operations should be dealt with internally of the memory allocator block (which means creation, allocation and release). Otherwise we'd have risks of leaks again.

Note the allocation of the buffer is not passed in, but dependent on the range length, through another call (`obtain memory`). Instead, it's not a buffer that is passed in, but a memory allocator object (see (here)[https://github.com/adobe/node-httptransfer/blob/b4f91c97100322faa6b9b3deac1b899136886b9b/lib/randomfileaccess.js#L94]). The allocation and slicing of buffer parts is encapsulated in the memory allocator and its helping private classes. Unfortunately, only code "outside" of the memory allocator knows when we're done with the access to a file part, and hence is the only one which knows when to release. Since this uses a slice of a larger backing buffer, I need to have that slice explicitly released to be able to return to "usable" memory (or implement more complex algortihms - but having a release function here seemed to be a good compromise). That is why that call is here.

Another reason to pass in the whole memory allocator is so that we can use the same instance of the memory allocator and underlying buffer in the up-/down-load pipelines. I didn't go for a singleton here because I think it may be useful to be able to have different instances of the allocator (and different backing buffers).